src/szk.rs

```diff
@@ -0,0 +1,469 @@
+    Decide(String),
+
+    /// Generic invalid parameter. This may be returned when a type cannot be constructed.
+    #[error("invalid paramter: {0}")]
+    InvalidParameter(String),
+
+    /// Returned if an FLP operation encountered an error.
+    #[error("Flp error: {0}")]
+    Flp(#[from] FlpError),
+
+    /// PRNG error.
+    #[error("prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// Codec error.
+    #[error("codec error: {0}")]
+    Codec(#[from] CodecError),
+}
+
+/// Contains an FLP proof share, and if joint randomness is needed, the blind
+/// used to derive it and the other party's joint randomness part.
+pub enum SzkProofShare<'a, const SEED_SIZE: usize, F: FftFriendlyFieldElement> {
+    ///Leader's proof share is uncompressed.
+    ///The first Seed is a blind, second is a joint randomness part.
+    Leader(Vec<F>, Option<&'a Seed<SEED_SIZE>>, Option<Seed<SEED_SIZE>>),
+    ///Helper's seed is compressed using the same blind as its joint randomness
+    Helper(&'a Seed<SEED_SIZE>, Option<Seed<SEED_SIZE>>),
+}
+
+/// A tuple of two proof shares: one for each aggregator
+pub type SzkProofShares<'a, const SEED_SIZE: usize, F> = (
+    SzkProofShare<'a, SEED_SIZE, F>,
+    SzkProofShare<'a, SEED_SIZE, F>,
+);
+/// A tuple containing the state and messages produced by an SZK query.
+pub type SzkQueryShares<const SEED_SIZE: usize, F> = (
+    Option<Seed<SEED_SIZE>>,
+    SzkVerifier<F>,
+    Option<Seed<SEED_SIZE>>,
+);
+/// Verifier type for Szk Proof
+pub type SzkVerifier<F> = Vec<F>;
+
+///Main struct encapsulating the shared zero-knowledge functionality.
+/// The type T is the underlying FLP proof system.
+/// P is the XOF used to derive all random coins (it should be indifferentiable
+/// from a random oracle for security.)
+#[cfg(test)]
+pub struct Szk<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    typ: T,
+    algorithm_id: u32,
+    phantom: PhantomData<P>,
+}
+
+#[cfg(test)]
+impl<T: Type> Szk<T, XofTurboShake128, 16> {
+    /// Create an instance of [`Szk`] using [`XofTurboShake128`]. The caller provides the bit length of
+    /// each measurement.
+    pub fn new_turboshake128(typ: T, algorithm_id: u32) -> Self {
+        Szk::new(typ, algorithm_id)
+    }
+}
+
+#[cfg(test)]
+impl<T, P, const SEED_SIZE: usize> Szk<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Construct an instance of this sharedZK proof system with the underlying FLP.
+    pub fn new(typ: T, algorithm_id: u32) -> Self {
+        Self {
+            typ,
+            algorithm_id,
+            phantom: PhantomData,
+        }
+    }
+
+    /// The output length of the underlying FLP.
+    pub fn output_len(&self) -> usize {
+        self.typ.output_len()
+    }
+
+    /// The verifier length of the underlying FLP.
+    pub fn verifier_len(&self) -> usize {
+        self.typ.verifier_len()
+    }
+    #[cfg(test)]
+    fn adjust_domsep_tag(&self, usage: u16) -> [u8; 8] {
```

Comment:
It looks like `usage` must be one of the `DST_` constants above, but this isn't enforced. I think this would be more clear with an `enum Usage`, because then you'd only define variants for usages that exist.

Reply:
As above, this gets a little complicated since there might be other usages in the same VDAF that SZK is not aware of.